File: implementations/kotlin/oshash.kt
// OpenSubtitles Hash (OSHash) - Kotlin implementation.
import java.io.File
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import kotlin.math.max
import kotlin.math.min

const val HASH_CHUNK_SIZE = 64 * 1024

fun computeHash(file: File): String {
    val size = file.length()
    val chunkSizeForFile = min(HASH_CHUNK_SIZE.toLong(), size)

    FileInputStream(file).channel.use { fileChannel ->
        val head = computeChunkHash(
            fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, chunkSizeForFile))
        val tail = computeChunkHash(
            fileChannel.map(FileChannel.MapMode.READ_ONLY, max(size - HASH_CHUNK_SIZE, 0), chunkSizeForFile))
        return String.format("%016x", size + head + tail)
    }
}

fun computeChunkHash(buffer: ByteBuffer): Long {
    val longBuffer = buffer.order(ByteOrder.LITTLE_ENDIAN).asLongBuffer()
    var hash = 0L
    while (longBuffer.hasRemaining()) {
        hash += longBuffer.get()
    }
    return hash
}

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        System.err.println("Usage: kotlin OshashKt <file>")
        System.exit(1)
    }
    println(computeHash(File(args[0])))
}

File: implementations/clojure/oshash.clj
;; OpenSubtitles Hash (OSHash) - Clojure implementation.
;; Algorithm: file_size + sum_le_u64(first_64KB) + sum_le_u64(last_64KB)

(import '[java.io RandomAccessFile]
        '[java.nio ByteBuffer ByteOrder])

(def chunk-size 65536)

(defn compute-hash [filepath]
  (let [raf (RandomAccessFile. filepath "r")
        fsize (.length raf)
        buf (byte-array chunk-size)]
    ;; Read first 64KB
    (.seek raf 0)
    (.readFully raf buf)
    (let [bb1 (-> (ByteBuffer/wrap buf) (.order ByteOrder/LITTLE_ENDIAN) .asLongBuffer)
          head (loop [i 0 h 0]
                 (if (< i (/ chunk-size 8))
                   (recur (inc i) (unchecked-add h (.get bb1)))
                   h))
          ;; Read last 64KB
          _ (.seek raf (max 0 (- fsize chunk-size)))
          _ (.readFully raf buf)
          bb2 (-> (ByteBuffer/wrap buf) (.order ByteOrder/LITTLE_ENDIAN) .asLongBuffer)
          tail (loop [i 0 h 0]
                 (if (< i (/ chunk-size 8))
                   (recur (inc i) (unchecked-add h (.get bb2)))
                   h))]
      (.close raf)
      (format "%016x" (unchecked-add fsize (unchecked-add head tail))))))

(when-let [filepath (first *command-line-args*)]
  (println (compute-hash filepath)))
